### PR TITLE
feat(cargo): support range=update-lockfile

### PR DESCRIPTION
**Pull request by rami3l**
## Changes

Adds `update-lockfile` rangeStrategy support to the `cargo` manager.

## Context

As requested in https://github.com/renovatebot/renovate/issues/22820.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

---

Hi there, and thanks a lot for making this tool!
I'm from WG-Rustup of the Rust team and we'd like to have this feature to simplify Rustup's regular maintenance process. 
I'm not very experienced with TypeScript, but I do want to drive #23968 thru the finish line.
Thanks in advance for your help!

### CODE REVIEW

File: lib/modules/manager/cargo/extract.ts [hunk 1]
```diff
@@ -286,9 +289,41 @@ export async function extractPackageFile(
...
+      const packageName = dep.packageName ?? dep.depName!;
+      const versions = coerceArray(versionsByPackage.get(packageName));
+      try {
+        const lockedVersion = versioning.getSatisfyingVersion(
+          versions,
+          dep.currentValue!
+        );
+        if (lockedVersion) {
+          dep.lockedVersion = lockedVersion;
+        }
```
Comment (rarkins) [inline lib/modules/manager/cargo/extract.ts:317]: Should we log a debug message if there's an "else"? Wouldn't it mean that there's no locked version for a dependency?

File: lib/modules/manager/cargo/artifacts.ts [hunk 2]
```diff
@@ -33,24 +36,53 @@ async function cargoUpdate(
...
+  };
+
+  await exec(cmds, execOptions);
+}
+
+export async function updateArtifacts(
+  updateArtifact: UpdateArtifact,
+): Promise<UpdateArtifactsResult[] | null> {
+  return await updateArtifactsImpl(updateArtifact);
+}
```
Comment (rarkins) [inline lib/modules/manager/cargo/artifacts.ts:73]: Can this refactoring be split into a separate PR to be merged first?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/modules/manager/cargo/index.ts [hunk 3]
```diff
@@ -4,10 +4,11 @@ import * as cargoVersioning from '../../versioning/cargo';
 import { updateArtifacts } from './artifacts';
 import { extractPackageFile } from './extract';
 export { getRangeStrategy } from './range';
+import { updateLockedDependency } from './update-locked';
```
Comment (viceice) [inline lib/modules/manager/cargo/index.ts:7]:
```suggestion
export { updateLockedDependency } from './update-locked';
```

File: lib/modules/manager/cargo/index.ts [hunk 4]
```diff
@@ -4,10 +4,11 @@ import * as cargoVersioning from '../../versioning/cargo';
...
 
 export const supportsLockFileMaintenance = true;
 
-export { extractPackageFile, updateArtifacts };
+export { extractPackageFile, updateArtifacts, updateLockedDependency };
```
Comment (viceice) [inline lib/modules/manager/cargo/index.ts:11]:
```suggestion
export { extractPackageFile, updateArtifacts };
```

File: lib/modules/manager/cargo/artifacts.ts (hunk 2 above)
Comment (rami3l) [reply lib/modules/manager/cargo/artifacts.ts:73]: Sorry, I'm not sure I have correctly understood your intentions. Isn't this one of the suggestions being made to the original patch https://github.com/renovatebot/renovate/pull/23968#discussion_r1382872136?

Comment (rarkins) [reply lib/modules/manager/cargo/artifacts.ts:73]: What I mean is that: you have both added a feature, but also done some refactoring of existing code in order to introduce it. We prefer to split refactoring and feature addition so that it's easier to review. e.g. the first PR is refactor-only, should require little or no changes of tests, etc. Then the feature is added on top, and easier to review on its own.

Comment (rami3l) [reply lib/modules/manager/cargo/artifacts.ts:73]: @rarkins Fair enough. I have made https://github.com/renovatebot/renovate/pull/26009 for this, hope this will help with your workflow.

File: lib/modules/manager/cargo/extract.spec.ts [hunk 5]
```diff
@@ -15,6 +15,12 @@ const cargo4toml = Fixtures.get('Cargo.4.toml');
 const cargo5toml = Fixtures.get('Cargo.5.toml');
 const cargo6configtoml = Fixtures.get('cargo.6.config.toml');
 const cargo6toml = Fixtures.get('Cargo.6.toml');
+const cargo7toml = Fixtures.get('Cargo.7.toml');
+const cargo7lock = Fixtures.get('Cargo.7.lock');
+const cargo8toml = Fixtures.get('Cargo.8.toml');
+const cargo8lock = Fixtures.get('Cargo.8.lock');
+const cargo9toml = Fixtures.get('Cargo.9.toml');
+const cargo9lock = Fixtures.get('Cargo.9.lock');
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.spec.ts:23]: do we really need so many new fixtures? can't we reuse existing or update existing fixtures?

File: lib/modules/manager/cargo/extract.ts [hunk 6]
```diff
@@ -286,9 +289,45 @@ export async function extractPackageFile(
...
+        );
+        if (lockedVersion) {
+          dep.lockedVersion = lockedVersion;
+        } else {
+          logger.debug(
+            `No locked version found for package ${dep.depName} in the range of ${dep.currentValue}.`,
+          );
+        }
+      } catch (err) {
+        // istanbul ignore next: this should not happen
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.ts:323]: can this be triggered by a invalid version inside the toml file?

File: lib/modules/manager/cargo/extract.spec.ts (hunk 5 above)
Comment (rami3l) [reply lib/modules/manager/cargo/extract.spec.ts:23]: I have reduced the number of fixtures as much as possible and have moved them to a new folder to further reduce visual clutters.

File: lib/modules/manager/cargo/extract.ts (hunk 6 above)
Comment (rami3l) [reply lib/modules/manager/cargo/extract.ts:323]: @viceice No. If I have an invalid version inside the toml file, the code will enter the `else` branch at L318, and no `lockedVersion` will not be found in `res?.deps[0]`.

**Comment by rami3l** [comment]
Sorry, I had an oversight when removing extra fixtures in a63e0be190297929c6228b1b5fd924dc03fe79dc, but I just made 531b528917cc02b8b72fbcc14eccd938fc928fcb to fix that. Would you mind helping me restart the CI checks?

File: lib/modules/manager/cargo/extract.spec.ts [hunk 7]
```diff
@@ -6,6 +6,7 @@ import { GlobalConfig } from '../../../config/global';
 import type { RepoGlobalConfig } from '../../../config/types';
 import { writeLocalFile } from '../../../util/fs';
 import type { ExtractConfig } from '../types';
+
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.spec.ts:9]:
```suggestion
```

File: lib/modules/manager/cargo/extract.ts (hunk 6 above)
Comment (viceice) [reply lib/modules/manager/cargo/extract.ts:323]: i think we should remove the try-catch then and let throw any unexpected error to get more visibillity

File: lib/modules/manager/cargo/locked-version.spec.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        localDir: join(tmpDir.path, 'local'),
+      };
+
+      GlobalConfig.set(adminConfig);
+    });
+
+    afterEach(async () => {
+      await tmpDir.cleanup();
+      GlobalConfig.reset();
+    });
```
Comment (viceice) [inline lib/modules/manager/cargo/locked-version.spec.ts:26]: just mock `util/fs` instead. it's much simpler and faster than copying files around

**Comment by rami3l** [comment]
@viceice Thanks a lot for taking your time to review my patch! One friendly suggestion though:

Would you mind pointing out all the requested changes for the current status of the PR in one go?
This way it should save time both for your team reviewing various PRs and our team trying to resolve many other issues at the same time; also, this will prevent myself from feeling too overwhelmed by the new requested changes coming up over and over. 🙇

File: lib/modules/manager/cargo/extract.spec.ts (hunk 7 above)
Comment (rami3l) [reply lib/modules/manager/cargo/extract.spec.ts:9]: Sorry, this was a miscommit. Actually it is a mitigation of the following issue:

For the following VSCode config:

```json
 "editor.codeActionsOnSave": {
    "source.organizeImports": true
  },
```

The `... from '.'` line below will be reordered on save and cause a hard error.

File: lib/modules/manager/cargo/extract.ts (hunk 6 above)
Comment (rami3l) [reply lib/modules/manager/cargo/extract.ts:323]: Done.

File: lib/modules/manager/cargo/locked-version.spec.ts (hunk 8 above)
Comment (rami3l) [reply lib/modules/manager/cargo/locked-version.spec.ts:26]: @viceice A newbie question: could you please help me out with the mock?

I have used `jest.mock('../../../util/fs');` at the top of the file, and it works for all `util/fs` calls within this module, but not for transitive calls that also indirectly uses it.

For example, in

```ts
    it('extracts registry urls from .cargo/config.toml', async () => {
      await writeLocalFile('.cargo/config.toml', cargo6configtoml);
      const res = await extractPackageFile(cargo6toml, 'Cargo.toml', {
        ...config,
      });
      expect(res?.deps).toMatchSnapshot();
      expect(res?.deps).toHaveLength(3);
    });
```

The `extractPackageFile()` call (which depends on `readLocalFile()`) cannot correctly read the file written by `writeLocalFile()` in the previous line, which causes this test to fail.

Comment (rami3l) [reply lib/modules/manager/cargo/locked-version.spec.ts:26]: Oh, never mind, looks like I can use `fs.readLocalFile.mockResolvedValueOnce()` instead of `writeLocalFile()`.

File: lib/modules/manager/cargo/locked-version.spec.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import { fs } from '../../../../test/util';
+import { GlobalConfig } from '../../../config/global';
+import type { RepoGlobalConfig } from '../../../config/types';
+import { extractLockFileVersions, parseLockFile } from './locked-version';
+
+jest.mock('../../../util/fs');
+
+function mockReadLocalFile(files: Record<string, string | null>) {
+  fs.readLocalFile.mockImplementation((file): Promise<any> => {
+    let content: string | null = '';
```
Comment (viceice) [inline lib/modules/manager/cargo/locked-version.spec.ts:13]: why default value is empty string instead of null?

File: lib/modules/manager/cargo/locked-version.spec.ts [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  });
+}
+
+describe('modules/manager/cargo/locked-version', () => {
+  describe('extractLockFileVersions()', () => {
+    let adminConfig: RepoGlobalConfig;
+    let tmpDir: DirectoryResult;
+
+    beforeEach(async () => {
+      tmpDir = await dir({ unsafeCleanup: true });
```
Comment (viceice) [inline lib/modules/manager/cargo/locked-version.spec.ts:27]: do you still need this?

**Review by viceice — COMMENTED** [review]
nearly done 💪❤️🤗

File: lib/modules/manager/cargo/locked-version.spec.ts (hunk 10 above)
Comment (rami3l) [reply lib/modules/manager/cargo/locked-version.spec.ts]: Removed in c3d2a226587df35820d9be89d5fefab2c88f0f44.

File: lib/modules/manager/cargo/locked-version.spec.ts (hunk 9 above)
Comment (rami3l) [reply lib/modules/manager/cargo/locked-version.spec.ts]: Changed in 7a21fca0b8037590517802070b005d335e333429.

File: lib/modules/manager/cargo/extract.spec.ts [hunk 11]
```diff
@@ -513,5 +502,84 @@ replace-with = "mcorbin"
...
+      const cargolock = Fixtures.get('lockfile-update/Cargo.3.lock');
+      mockReadLocalFile({ 'Cargo.lock': cargolock });
+      fs.findLocalSiblingOrParent.mockResolvedValueOnce('Cargo.lock');
+
+      const cargotoml = codeBlock`[package]
+name = "test"
+version = "0.1.0"
+edition = "2021"
+[dependencies]
+syn = "2.0"`;
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.spec.ts:516]:
```suggestion
      const cargotoml = codeBlock`
        [package]
        name = "test"
        version = "0.1.0"
        edition = "2021"
        [dependencies]
        syn = "2.0"
        `;
```
`codeBlock` will drop leading whitespace for you 😉

File: lib/modules/manager/cargo/extract.spec.ts [hunk 12]
```diff
@@ -513,5 +502,84 @@ replace-with = "mcorbin"
...
+      const cargolock = Fixtures.get('lockfile-update/Cargo.3.lock');
+      mockReadLocalFile({ 'Cargo.lock': cargolock });
+      fs.findLocalSiblingOrParent.mockResolvedValueOnce('Cargo.lock');
+
+      const cargotoml = codeBlock`[package]
+name = "test"
+version = "0.1.0"
+edition = "2021"
+[dependencies]
+syn = "2.foo.1"`;
```
Comment (viceice) [inline lib/modules/manager/cargo/extract.spec.ts:564]:
```suggestion
      const cargotoml = codeBlock`
        [package]
        name = "test"
        version = "0.1.0"
        edition = "2021"
        [dependencies]
        syn = "2.foo.1"
        `;
```

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]
